add script to evaluate CSR SpMV kernel

adhithadias opened the pull request.

reikdas commented:
Can you add the CSR SpMV benchmarking to `eval.py`?

ulysses4ever commented on README.md:
```diff
@@ -65,3 +65,28 @@ python3 bench_sparse_cuda.py
+./tmp/csr-spmv /local/scratch/a/Suitesparse/heart1/heart1.mtx 1
+./tmp/csr-spmv /local/scratch/a/Suitesparse/heart1/heart1.mtx 2
+
+# edit the files in eval and skip lists in the file before running
+# benchmark_csr_spmv.py, if empty it will evaluate all matrices
+# to eval specific files add it to eval [] array without .mtx
+# to skip specific files add it to skip [] array without .mtx 
+# before executing the python script
+# this will save the execution times in benchmark_results/csr_spmv.csv file
```
I can't really parse what's written here. At least, adding full stops and commas could help. But maybe you could give this text another go.

adhithadias replied:
Removed this section because benchmark file was removed.

ulysses4ever commented on src/csr-spmv.cpp:
```diff
@@ -0,0 +1,254 @@
+    float *y = new float[csrMatrix.rows];
+
+    // calculate the matrix-vector product and time it
+    float* exec_time = new float[32];
```
As a rule of thumb, never put literal constants anywhere except in initialization of `const` variables.

reikdas replied:
I think we should move this `32` to a variable called `BENCHMARK_FREQ` or something (to make sure that we update both this and the loop end number below).

adhithadias replied:
I have added a variable with 31 instead of 32.

ulysses4ever commented on src/csr-spmv.cpp:
```diff
@@ -0,0 +1,254 @@
+    #pragma omp parallel for
+    #endif
+    for (int i = 0; i < csr.rows; i++) {
+        // y[i] = 0;
```
Drop this line?

adhithadias replied:
Removed.

ulysses4ever commented on src/csr-spmv.cpp:
```diff
@@ -0,0 +1,254 @@
+        }
+        // Skip comments (lines starting with '%')
+        else if (line[0] == '%') {
+            // std::cout << line << std::endl;
```
Drop?

adhithadias replied:
Removed.

adhithadias commented on README.md:
```diff
@@ -65,3 +65,28 @@ python3 bench_sparse_cuda.py
+mkdir -p tmp && g++ -DOPENMP -O3 src/csr-spmv.cpp -o tmp/csr-spmv -fopenmp
+
+# to compile for single-thread execution
+mkdir -p tmp && g++ -O3 src/csr-spmv.cpp -o tmp/csr-spmv -fopenmp
```
I am including the header file. Ideally, I could remove the header file using an ifdef. Would you like me to remove it?

ulysses4ever replied:
Oh, right. Since you already have a couple of ifdef OPENMP, I think, it makes sense to guard the heart file as well...

adhithadias replied:
Changed.

adhithadias commented:
> Can you add the CSR SpMV benchmarking to `eval.py`?

I added it to eval.py. **Do you want me to remove the benchmark file I have already added?** I think it's better to run it as a standalone benchmark. You only need to run it once and then you can use another Python script to combine the results.

ulysses4ever commented:
There's a good bunch of duplication across various bench* scripts already. Ideally, we'd want to have less of that, not more. I can see how it may be convenient to have one file per task. But it quickly turns into maintenance nightmare if you don't do a conscious job of de-duplicating and factoring out common parts from these... 

>  I think it's better to run it as a standalone benchmark.

Without separate scripts it could be achieved via command-line arguments or env vars.

reikdas commented on README.md:
```diff
@@ -65,3 +65,28 @@ python3 bench_sparse_cuda.py
+# to eval specific files add it to eval [] array without .mtx
+# to skip specific files add it to skip [] array without .mtx 
+# before executing the python script
+# this will save the execution times in benchmark_results/csr_spmv.csv file
```
Why don't you save the results in `results/` like all the other benchmarking scripts?

adhithadias replied:
Removed the benchmark file altogether. I did not realize there was a results directory. I have added a .csv file with single thread execution times of SpMV for all matrices.

reikdas commented on benchmark_scripts/benchmark_csr_spmv.py:
```diff
@@ -0,0 +1,83 @@
+BASE_PATH = os.path.join(FILEPATH)
+BASE_PATH = os.path.join(BASE_PATH, "..")
+
+MTX_DIR = "/local/scratch/a/Suitesparse/"
```
Nitpick, but I would prefer `os.path.join("/local", "scratch", "a", "Suitesparse")`. I know we don't always follow this, but we should try and use it when we remember to.

adhithadias replied:
Updated.

reikdas commented on eval.py:
```diff
@@ -12,13 +12,25 @@
+    compile_csr_spmv_code(core)
     with open(os.path.join(BASE_PATH, "results", "res.csv"), "w") as f:
-        f.write("Filename,SABLE,PSC,Speedup\n")
+        f.write("Filename,SABLE,PSC,CSRSpmv,Speedup\n")
```
Can we have 2 "Speedup" columns? One over PSC and one over CSRSpmv?

adhithadias replied:
This file is removed.

adhithadias replied:
Added another column called "SpeedupOverCSRSpmv", and changed the original "Speedup" to "SpeedupOverPSC". Let me know if we need to change the column name.

reikdas commented:
> Do you want me to remove the benchmark file I have already added?

I think we can remove it for now, and then I'll make a different PR to decouple benchmarks (for both PSC and this).

reikdas commented on src/csr-spmv.cpp:
```diff
@@ -0,0 +1,254 @@
+    float* exec_time = new float[32];
+    spmv(csrMatrix, x, y);
+    for (int i = 0; i < 32; i++) {
+        auto start = std::chrono::high_resolution_clock::now();
```
Can we use `gettimeofday` instead of `std::chrono` like we do in our generated code? It is probably best to use the same timer in as many places as possible.

adhithadias replied:
Changed to use timeofday.

adhithadias commented on benchmark_results/dummy.txt:
Removed the directory all together because I removed the benchmark.py file.

adhithadias commented on src/csr-spmv.cpp:
```diff
@@ -0,0 +1,254 @@
+    }
+};
+
+// Define CSR struct without using vectors
```
Using vectors are slow.

adhithadias commented:
> > Do you want me to remove the benchmark file I have already added?
> 
> I think we can remove it for now, and then I'll make a different PR to decouple benchmarks (for both PSC and this), once we are done with our initial experiments.

Ideally, we should time them separately. There is no point in coupling it with other executions. For example, eval.py depends on having an executable (coming from another script) for our codegen strategy.

reikdas commented:
Is this ready to be merged?

adhithadias commented:
> Is this ready to be merged? Or are you still planning to make some changes?

Yes, this is complete.
